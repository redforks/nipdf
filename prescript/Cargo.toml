[package]
name = "prescript"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
winnow = { version = "0.5", features = ["simd"] }
log = "0.4"
either = "1.9.0"
memchr = "2.5.0"
ascii85 = "0.2.1"
educe = "0.4.22"
thiserror = "1.0"
anyhow = "1.0"
hex = "0.4"

[dev-dependencies]
test-case = "3.1"
test-log = "0.2"
env_logger = "0.10"
