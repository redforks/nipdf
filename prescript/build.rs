//! Build script that convert `names` file located at workspace root directory,
//! to src/name/built_in_names.rs.
//!
//! Each line in `names` file is a PostScript name, line start with '#' is a comment,
//! empty line is ignored.
//! Sort the names, and generate the builtin_in_names.rs file

use std::{fs::read_to_string, io::Write, path::Path};

fn main() {
    let names_file = Path::join(Path::new(env!("CARGO_MANIFEST_DIR")), "../names");
    let dest_file = Path::join(
        Path::new(env!("CARGO_MANIFEST_DIR")),
        "src/name/built_in_names.rs",
    );

    // return if names_file is not newer than dest_file
    if let Ok(names_meta) = std::fs::metadata(&names_file) {
        if let Ok(dest_meta) = std::fs::metadata(&dest_file) {
            if names_meta.modified().unwrap() <= dest_meta.modified().unwrap() {
                return;
            }
        }
    }

    let names = read_to_string(names_file).unwrap();
    let mut names: Vec<_> = names
        .lines()
        .map(|s| s.trim())
        .filter(|w| !w.is_empty() && !w.starts_with('#'))
        .collect();
    names.sort_unstable();
    names.dedup();
    let mut file = std::fs::File::create(dest_file).unwrap();
    let n = names.len();
    writeln!(file, "pub(crate) static BUILT_IN_NAMES: [&str; {n}] = [").unwrap();
    for name in names {
        writeln!(file, "    {:?},", name).unwrap();
    }
    writeln!(file, "];").unwrap();
}
