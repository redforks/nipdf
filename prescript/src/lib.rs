pub(crate) mod machine;
pub(crate) mod parser;

mod encoding;
mod type1;
pub use encoding::Encoding;
use string_interner::{backend::BucketBackend, symbol::SymbolU16, StringInterner};
pub use type1::Font;

/// Symbol for .notdef glyph
pub const NOTDEF: &str = ".notdef";
pub type Name = SymbolU16;
pub type NameRegistry = StringInterner<BucketBackend<Name>>;
