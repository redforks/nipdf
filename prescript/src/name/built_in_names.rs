#[rustfmt::skip]
pub const BUILT_IN_NAMES: [&str; 7] = ["<<", ">>", "[", "]", "for", "if", ".notdef"];
