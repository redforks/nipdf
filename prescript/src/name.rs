//! Efficient way to store PostScript Name Value
//!
//! PostScript has many frequently used names, such as font glyph names,
//! operation names etc. Use `String`, `Vec<str>`, `Box<str>` needs a lot
//! of allocations, and `Rc<str>` complex code logic.
//!
//! This crate provides a efficient way to store PostScript Name Value.
//! `Name` type wraps a `Either` enum type, which is `Either<u16, Box<str>>`.
//! Predefine common used names, given each name a unique u16 number, and
//! other names stored in Box<str>. Static strings stored in a const array,
//! sorted by alphabetical order, so we can use binary search to find a name.
//!
//! Use `name!()` macro to resolve `&'static str` to Name at compile time,
//! no runtime cost.

use either::Either::{self, Left, Right};

mod built_in_names;
// Hack to use name!() macro
use crate as prescript;
use built_in_names::BUILT_IN_NAMES;
use prescript_macro::name;
use std::fmt::{Display, Formatter};

/// PostScript Name Value
/// Important: Always use name function and macro create Name, don't create
/// Name directly. The internal structure exposed for pattern match combined with `name!` macro:
///
/// match name {
///   name!("foo") => { ... }
///   name!("bar") => { ...}
///   _ => { ... }
/// }  
#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Name(pub Either<u16, Box<str>>);

impl std::fmt::Debug for Name {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "name!({})", self.as_ref())
    }
}

/// Special name to match normally won't exist.
pub static INVALID1: Name = name!("$$invalid1$$");
/// Special name to match normally won't exist.
pub static INVALID2: Name = name!("$$invalid2$$");

impl Display for Name {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "/{}", self.as_ref())
    }
}

impl AsRef<str> for Name {
    fn as_ref(&self) -> &str {
        match &self.0 {
            Left(i) => BUILT_IN_NAMES[*i as usize],
            Right(s) => s.as_ref(),
        }
    }
}

/// Create Name from `&str`, if it is one of builtin names, `Name` will
/// use `u16` to store it, otherwise, `Name` will use `Box<str>` to store it.
///
/// Preferred to use `name!()` macro if possible.
pub fn name(s: &str) -> Name {
    // binary search BUILT_IN_NAMES, return its index if found, otherwise return boxed str
    match BUILT_IN_NAMES.binary_search(&s) {
        Ok(i) => Name(Left(i as u16)),
        Err(_) => Name(Right(Box::from(s))),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use static_assertions::assert_eq_size;

    #[test]
    fn test_name_size() {
        assert_eq_size!(Name, (usize, usize));
    }

    #[test]
    fn name_from_str() {
        assert!(matches!(name("foo"), Name(Right(_))));
        assert!(matches!(name("for"), Name(Left(_))));
    }

    #[test]
    fn as_str() {
        assert_eq!("foo", name("foo").as_ref());
        assert_eq!("for", name("for").as_ref());
    }

    #[test]
    fn name_macro() {
        assert_eq!(name!("foo"), name("foo"));
        assert_eq!(name!("for"), name("for"));
    }

    #[test]
    fn display() {
        assert_eq!(format!("{}", name!("foo")), "/foo");
        assert_eq!(format!("{}", name!("for")), "/for");
    }
}
