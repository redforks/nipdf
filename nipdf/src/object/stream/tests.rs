use super::*;
use crate::{file::decode_stream, function::Domain, object::Name};
use prescript::name;
use test_case::test_case;

#[test_case([] => Ok(vec![]); "empty")]
#[test_case(
    [(&KEY_FILTER, 1.into())] => matches Err(ObjectValueError::UnexpectedType);
    "incorrect filter type"
)]
#[test_case(
    [(&KEY_FILTER, Object::Array(vec![1.into()]))] => matches Err(_);
    "filter is array but item not name"
)]
#[test_case(
    [(&KEY_FILTER, FILTER_FLATE_DECODE.clone().into())] =>
    Ok(vec![(FILTER_FLATE_DECODE.clone(), None)]);
     "one filter"
)]
#[test_case(
    [(&KEY_FILTER, FILTER_FLATE_DECODE.clone().into()),
     (&KEY_FILTER_PARAMS, Object::Null)] =>
    Ok(vec![(FILTER_FLATE_DECODE.clone(), None)]);
     "one filter with null params"
)]
#[test_case(
    [(&KEY_FILTER, FILTER_FLATE_DECODE.clone().into()),
     (&KEY_FILTER_PARAMS, Object::Array(vec![Object::Null]))] =>
    Ok(vec![(FILTER_FLATE_DECODE.clone(), None)]);
     "one filter with null params in array"
)]
#[test_case(
    [(&KEY_FILTER, FILTER_FLATE_DECODE.clone().into()),
     (&KEY_FILTER_PARAMS, Object::Dictionary(Dictionary::default()))] =>
    Ok(vec![(FILTER_FLATE_DECODE.clone(), Some(Dictionary::default()))]);
     "one filter with dictionary params"
)]
#[test_case(
    [(&KEY_FILTER, vec![
        FILTER_FLATE_DECODE.clone().into(),
        FILTER_DCT_DECODE.clone().into(),
    ].into())] =>
    Ok(vec![(FILTER_FLATE_DECODE.clone(), None),
            (FILTER_DCT_DECODE.clone(), None)]);
     "two filters no params"
)]
#[test_case(
    [(&KEY_FILTER, vec![
        FILTER_FLATE_DECODE.clone().into(),
        FILTER_DCT_DECODE.clone().into(),
    ].into()),
    (&KEY_FILTER_PARAMS, Dictionary::default().into())] =>
    Ok(vec![(FILTER_FLATE_DECODE.clone(), Some(Dictionary::default())),
            (FILTER_DCT_DECODE.clone(), None)]);
     "two filters with null params"
)]
#[test_case(
    [(&KEY_FFILTER, FILTER_FLATE_DECODE.clone().into())] =>
    Err(ObjectValueError::ExternalStreamNotSupported);
     "filter not supported"
)]
fn test_iter_filter(
    dict: impl IntoIterator<Item = (&'static Name, Object<'static>)>,
) -> Result<Vec<(Name, Option<Dictionary<'static>>)>, ObjectValueError> {
    let dict: Dictionary<'static> = dict
        .into_iter()
        .map(|(k, v)| (k.clone(), v))
        .collect::<Dictionary>();
    let stream = Stream(dict, &[], ObjectId::empty());
    let r: Vec<(Name, Option<Dictionary<'static>>)> = stream
        .iter_filter()?
        .map(|(k, v)| (k.clone(), v.cloned()))
        .collect();
    Ok(r)
}

#[test_case(228, 228, 227 => 228)]
#[test_case(0, 0, 1 => 0; "close to a")]
#[test_case(0, 0, 2 => 0; "close to b")]
#[test_case(0, 0, 3 => 0; "close to c")]
fn test_paeth(a: u8, b: u8, c: u8) -> u8 {
    paeth(a, b, c)
}

#[test]
fn predictor_8bit() {
    insta::assert_debug_snapshot!(
        &decode_stream("filters/predictor.pdf", 7u32, |d, resolver| {
            let params = d.get(&name!("DecodeParms")).unwrap().as_dict()?;
            assert_eq!(
                15,
                resolver
                    .resolve_container_value(params, name!("Predictor"))?
                    .as_int()?
            );
            Ok(())
        })
        .unwrap()[127..255]
    );
}

#[test]
fn predictor_24bit() {
    insta::assert_debug_snapshot!(
        &decode_stream("color-space/cal-rgb.pdf", 6u32, |d, resolver| {
            let params = d.get(&name!("DecodeParms")).unwrap().as_dict()?;
            assert_eq!(
                15,
                resolver
                    .resolve_container_value(params, name!("Predictor"))?
                    .as_int()?
            );
            Ok(())
        })
        .unwrap()[..255]
    );
}

#[test]
fn image_mask_try_from_object() {
    // ColorKeyMask
    #[rustfmt::skip]
    let o = Object::Array(
        vec![
            0.into(), 1.into(), // domain 1
            0.1.into(), 0.9.into(), // domain 2
            0.2.into(), 0.8.into(), // domain 3
        ],
    );
    let mask = ImageMask::try_from(&o).unwrap();
    assert_eq!(
        mask,
        ImageMask::ColorKey(Domains(vec![
            Domain::new(0., 1.),
            Domain::new(0.1, 0.9),
            Domain::new(0.2, 0.8),
        ]))
    );

    // ExplicitMask
    let stream = Stream(
        Dictionary::default(),
        b"0 1 2 3 4 5 6 7 8 9".as_ref(),
        ObjectId::empty(),
    );
    let o = Object::Stream(stream.clone());
    let mask = ImageMask::try_from(&o).unwrap();
    assert_eq!(mask, ImageMask::Explicit(stream));
}

#[test_case([10, 15, 20, 0] => true; "matches lower range")]
#[test_case([110, 115, 120, 0] => true; "matches upper range")]
#[test_case([10, 15, 20, 1] => true; "alpha not compared")]
#[test_case([109, 114, 120, 0] => true; "in range")]
#[test_case([9, 15, 20, 0] => false; "red part less than min")]
#[test_case([111, 115, 120, 0] => false; "red part greater than max")]
#[test_case([10, 14, 20, 0] => false; "green part less than min")]
#[test_case([110, 116, 120, 0] => false; "green part greater than max")]
fn test_color_matches_color_key(color: [u8; 4]) -> bool {
    let color_key: ColorKey = ([10, 15, 20, 0], [110, 115, 120, 0]);
    color_matches_color_key(color_key, color)
}

#[test]
fn test_color_key_range() {
    let range = Domains(vec![
        Domain::new(10., 110.),
        Domain::new(15., 115.),
        Domain::new(20., 120.),
    ]);
    let color_key: ColorKey = ([10, 15, 20, 255], [110, 115, 120, 255]);
    assert_eq!(color_key, color_key_range(&range, &ColorSpace::DeviceRGB));
}
