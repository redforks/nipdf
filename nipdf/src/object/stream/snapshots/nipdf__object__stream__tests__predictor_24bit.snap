---
source: nipdf/src/object/stream/tests.rs
expression: "&decode_stream(\"color-space/cal-rgb.pdf\", 6u32,\n                |d, resolver|\n                    {\n                        let params = d.get(&b\"DecodeParms\"[..]).unwrap().as_dict()?;\n                        assert_eq!(15,\n                        resolver.resolve_container_value(params, \"Predictor\")\n                        ?.as_int() ?);\n                        Ok(())\n                    }).unwrap()[..255]"
---
[
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
    255,
    0,
    0,
]
